soroban-env-host/src/auth.rs

```diff
@@ -0,0 +1,695 @@
+            // needs to be authorized) or B, C (in case if A call doesn't need
+            // to be authorized), but not B, A, A->C (if authorization on A has
+            // been called after calling B).
+            if authorized_for_curr_frame.is_none() {
+                *authorized_for_curr_frame = Some(false);
+            }
+        }
+        self.authorized_for_frame.push(None);
+    }
+
+    fn pop_frame(&mut self) {
+        self.authorized_for_frame.pop();
+        // Require nonce for the root contract of every independent call tree.
+        let has_authorized_frames = self
+            .authorized_for_frame
+            .iter()
+            .find(|f| f == &&Some(true))
+            .is_some();
+        if !has_authorized_frames {
+            self.need_nonce = true;
+        }
+    }
+
+    fn mark_frame_authorization(&mut self) -> Result<(), HostError> {
+        if let Some(authorized_for_curr_frame) = self.authorized_for_frame.last_mut() {
+            if authorized_for_curr_frame.is_none() {
+                *authorized_for_curr_frame = Some(true);
+                Ok(())
+            } else {
+                // Maybe we should allow to call authorize multiple times, but
+                // only the first call would actually be checked/recorded...
+                Err(ScUnknownErrorCode::General.into())
+            }
+        } else {
+            // This would be an internal error, maybe should panic instead
+            Err(ScUnknownErrorCode::General.into())
+        }
+    }
+
+    fn get_authorized_call_stack(
+        &self,
+        call_stack: &[ContractInvocation],
+    ) -> Result<Vec<ContractInvocation>, HostError> {
+        let mut res = vec![];
+        if call_stack.len() != self.authorized_for_frame.len() {
+            // This would be an internal error, maybe should panic instead
+            return Err(ScUnknownErrorCode::General.into());
+        }
+        for i in 0..self.authorized_for_frame.len() {
+            if self.authorized_for_frame[i] == Some(true) {
+                res.push(call_stack[i].clone());
+            }
+        }
+        Ok(res)
+    }
+
+    fn maybe_consume_nonce(
+        &mut self,
+        host: &Host,
+        top_invocation: &ContractInvocation,
+    ) -> Result<Option<u64>, HostError> {
+        if !self.need_nonce {
+            return Ok(None);
+        }
+        self.need_nonce = false;
+        Ok(Some(host.read_and_consume_nonce(
+            &top_invocation.contract_id,
+            &self.address,
+        )?))
+    }
+
+    fn authorize_invocation(
+        &mut self,
+        host: &Host,
+        call_stack: &[ContractInvocation],
+        top_args: ScVec,
+    ) -> Result<(), HostError> {
+        self.maybe_authenticate(host)?;
+        self.mark_frame_authorization()?;
+        let nonce = self.maybe_consume_nonce(host, &call_stack[0])?;
+        let expected_invocation = AuthorizedInvocation::new(
+            self.get_authorized_call_stack(call_stack)?,
```

Comment:
I have a question related to the authorized call stack. Let's say I have invocations A -> B -> C, where C calls `account.authorize(args)`. Am I right in saying that `get_authorized_call_stack` will only return the C call? How do you get it to return A -> B -> C? By calling `authorize` in each contract? Just trying to understand the details of how the call stack is managed.

Reply:
If only C requires authorization then anyone can call it. Basically the idea is that every contract should decide whether it needs authorization or not. The alternative would be to require every call to be signed, but after the discussions it seems like that would be too limiting (e.g. see my multiswap example).